Adds Android foreground location support

Wackymax opened the pull request:
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)

Feature: Adds support for background (Foreground service) support to Android for when the applications goes into the background while location stream is active

Related to PR: https://github.com/Baseflow/flutter-geolocator/pull/957

### :arrow_heading_down: What is the current behavior?

Currently when the application goes into background and the OS kills the activity then location streaming stops.

### :new: What is the new behavior (if this is a feature change)?

New options have been introduced on Android that will allow the users to configure the background behaviour allow the service to continue running in the background keeping the activity alive while the location stream is active.

### :boom: Does this PR introduce a breaking change?

A new Android service is introduced which changes the behaviour of the location stream to use the new service but existing apps will not know about this change.

### :bug: Recommendations for testing

Testing the normal mode to validate the location stream is still working as expected and testing the background mode with the application running in the background.

### :memo: Links to relevant issues/docs

https://github.com/Baseflow/flutter-geolocator/issues/948
https://github.com/Baseflow/flutter-geolocator/issues/53

### :thinking: Checklist before submitting

- [X] I made sure all projects build.
- [X] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy](https://dart.dev/tools/pub/versioning).
- [X] I updated CHANGELOG.md to add a description of the change.
- [X] I followed the style guide lines ([code style guide](https://github.com/Baseflow/flutter-geolocator/blob/master/CONTRIBUTING.md)).
- [X] I updated the relevant documentation.
- [X] I rebased onto current `master`.

mvanbeusekom commented on geolocator_android/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+## 3.1.0
+
+- Added a new option to Android settings that will allow us to configure the foreground service notification settings
+- Added wake lock and WiFI lock capabilities to the Android service
+- Added an Android service to retrieve location updates. This allows us to make the service a foreground service which will stay alive and continue processing location updates if if the application is moved into the background.
```
Nit: We should convert this into one bullet point as the current bullet points describe the implementation and release notes normally don't list implementation detail. I think we could convert this into something like:

```suggestion
- Adds support to make a foreground service and continue processing location updates when the application is moved into the background.
```

mvanbeusekom commented on geolocator_android/CHANGELOG.md:
```diff
@@ -61,4 +67,3 @@
 ## 1.0.0
 
 - Initial open source release.
-
```
Nit: Please make sure all files end with an empty line. This is to ensure (older) POSIX

mvanbeusekom commented on geolocator_android/android/src/main/java/com/baseflow/geolocator/MethodCallHandlerImpl.java:
```diff
@@ -124,7 +126,11 @@ void setActivity(@Nullable Activity activity) {
   }
 
-  private void onCheckPermission(MethodChannel.Result result) {
+    void setBackgroundLocationService(@Nullable GeolocatorLocationService backgroundLocationService) {
```
Rename this too `setForegroundLocationService` and the parameter too `foregroundLocationService` to remove confusion with the Android Background service feature.

mvanbeusekom commented on geolocator_android/android/src/main/java/com/baseflow/geolocator/StreamHandlerImpl.java:
```diff
@@ -21,24 +18,21 @@
-  void setActivity(@Nullable Activity activity) {
-    this.activity = activity;
-  }
+    public void setBackgroundLocationService(@Nullable GeolocatorLocationService backgroundLocationService) {
```
Same, rename this to `setForegroundLocationService` and the parameter to `foregroundLocationService`.

mvanbeusekom commented on geolocator_android/android/src/main/java/com/baseflow/geolocator/StreamHandlerImpl.java:
```diff
@@ -21,24 +18,21 @@
   @Nullable private Context context;
-  @Nullable private Activity activity;
-  @Nullable private LocationClient locationClient;
+    @Nullable private GeolocatorLocationService backgroundLocationService;
```
Alignment and rename to `foregroundLocationService`.

mvanbeusekom commented on geolocator_android/android/src/main/java/com/baseflow/geolocator/location/AndroidIconResource.java:
```diff
@@ -0,0 +1,34 @@
+    public String getDefType() {
+        return defType;
+    }
+}
```
Nit: Add missing empty line.

mvanbeusekom commented on geolocator_android/android/src/main/java/com/baseflow/geolocator/location/BackgroundNotification.java:
```diff
@@ -0,0 +1,102 @@
+    public Notification build() {
+        return builder.build();
+    }
+}
```
Nit: add missing empty line.

mvanbeusekom commented on geolocator_android/android/src/main/java/com/baseflow/geolocator/location/BackgroundNotification.java:
```diff
@@ -0,0 +1,102 @@
+            intent.setPackage(null);
+            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED);
+            int flags = PendingIntent.FLAG_UPDATE_CURRENT;
+            if (Build.VERSION.SDK_INT > 23) {
```
Use `Build.VERSION_CODES.M` instead to the `23` number literal:

```suggestion
            if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) {
```

mvanbeusekom commented on geolocator_android/android/src/main/java/com/baseflow/geolocator/location/ForegroundNotificationOptions.java:
```diff
@@ -0,0 +1,64 @@
+@SuppressWarnings({"ConstantConditions", "unchecked"})
+public class ForegroundNotificationOptions {
+
+    private final String notificationTitle;
+    private final String notificationText;
+    private final AndroidIconResource notificationIcon;
+    private final boolean enableWifiLock;
+    private final boolean enableWakeLock;
```
Please decorate with attributes to indicate nullability.

mvanbeusekom commented on geolocator_android/android/src/main/java/com/baseflow/geolocator/location/ForegroundNotificationOptions.java:
```diff
@@ -0,0 +1,64 @@
+    private final boolean enableWifiLock;
+    private final boolean enableWakeLock;
+
+
```
Nit: remove redundant empty line.

mvanbeusekom commented on geolocator_android/android/src/main/java/com/baseflow/geolocator/location/NotificationImportance.java:
```diff
@@ -0,0 +1,7 @@
+    Default,
+    High,
+    Max,
+}
```
Nit: add missing empty line.

mvanbeusekom commented on geolocator_android/lib/geolocator_android.dart:
```diff
@@ -1 +1,2 @@
-export 'src/types/android_settings.dart';
+export 'src/types/android_settings.dart' show AndroidSettings;
+export 'src/types/foreground_settings.dart' show ForegroundNotificationConfig;
```
Nit: add missing empty line.

mvanbeusekom commented on geolocator_android/lib/src/types/android_settings.dart:
```diff
@@ -34,12 +37,17 @@ class AndroidSettings extends LocationSettings {
+        'foregroundNotificationConfig': foregroundNotificationConfig?.toJson(),
       });
   }
 }
```
Nit: since you are modifying this file, it would be great if you could also add an empty line here.

mvanbeusekom commented on geolocator_android/lib/src/types/android_settings.dart:
```diff
@@ -34,12 +37,17 @@ class AndroidSettings extends LocationSettings {
   final Duration? intervalDuration;
 
+  /// If this is set then the services is started as a Foreground service with a persistent notification
+  /// showing the user that the service will continue running in the background
```
Nit: sentence should end with a period.
```suggestion
  /// showing the user that the service will continue running in the background.
```

mvanbeusekom commented on geolocator_android/lib/src/types/foreground_settings.dart:
```diff
@@ -0,0 +1,74 @@
+/// Uniquely identifies an Android resource
```
Nit: sentence should end with a period.
```suggestion
/// Uniquely identifies an Android resource.
```

mvanbeusekom commented on geolocator_android/lib/src/types/foreground_settings.dart:
```diff
@@ -0,0 +1,74 @@
+  /// Optional default resource type to find, if "type/" is not included in the name. Can be null to require an explicit type.
+  final String defType;
+
+  /// Uniquely identifies an Android resource
```
Nit: sentence should end with a period.
```suggestion
  /// Uniquely identifies an Android resource.
```

mvanbeusekom commented on geolocator_android/lib/src/types/foreground_settings.dart:
```diff
@@ -0,0 +1,74 @@
+  /// Uniquely identifies an Android resource
+  const AndroidResource({required this.name, this.defType = 'drawable'});
+
+  /// Returns a json representation of this class
```
Nit: sentence should end with a period and JSON should be all capitals.
```suggestion
  /// Returns a JSON representation of this class.
```

mvanbeusekom commented on geolocator_android/lib/src/types/foreground_settings.dart:
```diff
@@ -0,0 +1,74 @@
+  /// user has not used the device in a while (e.g. for background network
+  /// communications).
+  ///
+  /// Wifilock permissions should be obtained first by using a permissions library
```
Nit: sentence should end with a period.
```suggestion
  /// Wifi lock permissions should be obtained first by using a permissions library.
```

mvanbeusekom commented on geolocator_android/lib/src/types/foreground_settings.dart:
```diff
@@ -0,0 +1,74 @@
+
+  /// When enabled, a Wakelock is acquired when background execution is started.
+  ///
+  /// Wakelock permissions should be obtained first by using a permissions library
```
Nit: sentence should end with a period.

```suggestion
  /// Wake lock permissions should be obtained first by using a permissions library.
```

mvanbeusekom commented on geolocator_android/lib/src/types/foreground_settings.dart:
```diff
@@ -0,0 +1,74 @@
+    this.enableWakeLock = false,
+  });
+
+  /// Returns a json representation of this class
```
Nit: sentence should end with a period and JSON should be all capitals.
```suggestion
  /// Returns a JSON representation of this class.
```

mvanbeusekom commented on geolocator_android/pubspec.yaml:
```diff
@@ -18,7 +18,7 @@ flutter:
   flutter:
     sdk: flutter
-  geolocator_platform_interface: ^3.0.0+1
+  geolocator_platform_interface: ^4.0.0+1
```
Version `4.0.0+1` doesn't exists, please update to `4.0.3`.
```suggestion
  geolocator_platform_interface: ^4.0.3
```

mvanbeusekom reviewed (changes requested):
Thank you for splitting up the initial PR. I have reviewed these changes and the code overall looks very good. I just left some small remarks and some style change request. 

I would really appreciate it if you could go through these and provide an update.

Wackymax commented on geolocator_android/android/build.gradle:
```diff
@@ -43,4 +43,5 @@ android {
 dependencies {
     implementation 'com.google.android.gms:play-services-location:17.0.0'
+    implementation 'androidx.core:core:1.2.0'
 }
```
For some reason I already have an empty line on my side

mvanbeusekom replied:
You are absolutely right, apologies. GitHub doesn't show it indeed. Only when it is missing GitHub will show an extra warning. Please ignore my requests about this.

Wackymax commented on geolocator_android/android/src/main/java/com/baseflow/geolocator/GeolocatorLocationService.java:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    public void stopLocationService() {
+        if (this.locationClient != null && geolocationManager != null) {
```
There is a lot of code form the library that is using "this" in some places. I've made updates to remove as far as possible.

Wackymax commented on geolocator_android/android/src/main/AndroidManifest.xml:
```diff
@@ -1,3 +1,13 @@
+                android:foregroundServiceType="location"
+                android:name=".GeolocatorLocationService"/>
+    </application>
 </manifest>
```
Once again, I see a empty line on my side. Not sure where in the process it gets removed

Wackymax commented:
I've applied the PR suggestions that were made

Wackymax commented on geolocator_android/android/src/main/java/com/baseflow/geolocator/location/BackgroundNotification.java:
```diff
@@ -0,0 +1,102 @@
+import androidx.core.app.NotificationManagerCompat;
+
+public class BackgroundNotification {
+    private Context context;
+    private Integer notificationId;
+    private String channelId;
+    private ForegroundNotificationOptions options;
+    private NotificationCompat.Builder builder;
```
None of these can be null. Options wasn't being used so I removed it.

mvanbeusekom reviewed (commented):
@Wackymax thanks for applying the requested changes. 

Having another look at the code I noticed the example App doesn't contain any changes. I am not to familiar with foreground services so I am not to sure if it is required to do any changes to make use of the background services.

However there is a new `ForegroundNotificationConfig` class as part of the Dart code, I think it would be good to reflect its usage in the example App. Would you mind giving this an update?

P.S. sorry for not realizing this in the first round. Really appreciate your efforts.

mvanbeusekom commented on geolocator_android/lib/src/types/foreground_settings.dart:
```diff
@@ -38,6 +38,7 @@ class ForegroundNotificationConfig {
   /// When enabled, a Wakelock is acquired when background execution is started.
+  /// If this is false then the system can still sleep and all location events will be received at once when the system wakes up again.
   ///
   /// Wake lock permissions should be obtained first by using a permissions library.
```
According to the [Dart doc specifications](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#dartdoc-specific-requirements) the first paragraph should be a short self contained sentence followed by more elaborate paragraphs. Also lines should be wrapped at 80 characters according to the Dart style-guide.

Lets restructure this a little bit to for example:
```
  /// When enabled, a Wakelock is acquired when background execution is 
  /// started.
  ///
  /// If this is false then the system can still sleep and all location 
  /// events will be received at once when the system wakes up again.
  ///
  /// Wake lock permissions should be obtained first by using a permissions library.
```

Wackymax replied:
Done. Sorry for all the back and forth. This is my first contribution to an open source project 😅

mvanbeusekom reviewed (approved):
LGTM, if you can fix the formatting and the Dart doc for the `enableWakeLock` field (see additional comment).

mvanbeusekom commented:
Thank you for all your effort @Wackymax. The geolocator_android package has just been published, this means it is now also possible to update #980 with the new reference. It would also be good to update the example App here.

Wackymax commented:
Thank you @mvanbeusekom,

I've updated [980](https://github.com/Baseflow/flutter-geolocator/pull/980). The example app does not set any settings at all so it didn't feel right updating it with Android settings. But I added the example to the readme docs.
